backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/s3/S3Service.java

```diff
@@ -158,13 +177,27 @@ private void initializeS3ClientWithTemporaryCredentials(Regions clientRegion) {
 
   private void RefreshCredentialsAndSchedule(String region) {
     LOGGER.debug("fetching token for s3 access");
+    if (task != null) {
+      task.cancel();
+      task = null;
+    }
 
     LOGGER.debug(
         "credentials before refresh {}",
         temporarySessionCredentials != null ? temporarySessionCredentials.hashCode() : null);
 
-    TimerTask task = new FetchTemporaryS3Token(region);
-    task.run();
+    scheduleTimer(region);
+  }
+
+  public static synchronized void scheduleTimer(String region) {
+    if (task == null) {
+      task = new FetchTemporaryS3Token(region);
+      task.run();
+      return;
```

Comment:
Does `.run()` perform immediately? If not, this will return without having finished running, right?

Reply:
yes it calls https://github.com/VertaAI/modeldb/blob/master/backend/src/main/java/ai/verta/modeldb/cron_jobs/FetchTemporaryS3Token.java#L34 synchronously

Reply:
on a closer look this does look weird , @ad-47 why are we not scheduling the next run here?